packages/aws-cdk-lib/aws-cloudfront-origins/lib/s3-bucket-origin.ts

```diff
@@ -0,0 +1,251 @@
+import { Construct } from 'constructs';
+import * as cloudfront from '../../aws-cloudfront';
+import { AccessLevel } from '../../aws-cloudfront';
+import * as iam from '../../aws-iam';
+import { IKey } from '../../aws-kms';
+import { IBucket } from '../../aws-s3';
+import { Annotations, Aws, Names, Stack } from '../../core';
+
+const BUCKET_ACTIONS: Record<string, string[]> = {
+  READ: ['s3:GetObject'],
+  WRITE: ['s3:PutObject'],
+  DELETE: ['s3:DeleteObject'],
+};
+
+const KEY_ACTIONS: Record<string, string[]> = {
+  READ: ['kms:Decrypt'],
+  WRITE: ['kms:Encrypt', 'kms:GenerateDataKey*'],
+};
+
+/**
+ * Properties for configuring a origin using a standard S3 bucket
+ */
+export interface S3BucketOriginBaseProps extends cloudfront.OriginProps { }
+
+/**
+ * Properties for configuring a S3 origin with OAC
+ */
+export interface S3BucketOriginWithOACProps extends S3BucketOriginBaseProps {
+  /**
+  * An optional Origin Access Control
+  *
+  * @default - an Origin Access Control will be created.
+  */
+  readonly originAccessControl?: cloudfront.IOriginAccessControl;
+
+  /**
+   * The level of permissions granted in the bucket policy and key policy (if applicable)
+   * to the CloudFront distribution.
+   *
+   * @default [AccessLevel.READ]
+   */
+  readonly originAccessLevels?: AccessLevel[];
+}
+
+/**
+ * Properties for configuring a S3 origin with OAI
+ */
+export interface S3BucketOriginWithOAIProps extends S3BucketOriginBaseProps {
+  /**
+  * An optional Origin Access Identity
+  *
+  * @default - an Origin Access Identity will be created.
+  */
+  readonly originAccessIdentity?: cloudfront.IOriginAccessIdentity;
+}
+
+/**
+ * A S3 Bucket Origin
+ */
+export abstract class S3BucketOrigin extends cloudfront.OriginBase {
+  /**
+   * Create a S3 Origin with Origin Access Control (OAC) configured
+   */
+  public static withOriginAccessControl(bucket: IBucket, props?: S3BucketOriginWithOACProps): cloudfront.IOrigin {
+    return new class extends S3BucketOrigin {
+      private originAccessControl?: cloudfront.IOriginAccessControl;
+
+      constructor() {
+        super(bucket, { ...props });
+        this.originAccessControl = props?.originAccessControl;
+      }
+
+      public bind(scope: Construct, options: cloudfront.OriginBindOptions): cloudfront.OriginBindConfig {
+        if (!this.originAccessControl) {
+          this.originAccessControl = new cloudfront.S3OriginAccessControl(scope, 'S3OriginAccessControl');
+        }
+
+        const distributionId = options.distributionId;
+        const accessLevels = new Set(props?.originAccessLevels ?? [cloudfront.AccessLevel.READ]);
+        const bucketPolicyActions = this.getBucketPolicyActions(accessLevels);
+        const bucketPolicyResult = this.grantDistributionAccessToBucket(distributionId!, bucketPolicyActions);
```

Comment:
It will not throw an error but it will be an non-functional bucket policy statement since `undefined` cannot be a valid distribution ID

Reply:
I think it will throw an error for `undefined!`. However, seeing [where this is triggered ](https://github.com/aws/aws-cdk/blob/gracelu0/s3-oac-l2/packages/aws-cdk-lib/aws-cloudfront/lib/distribution.ts#L610), I think we're pretty safe to say that it will always be defined.